bpf: initial bpf-based masquerading support

borkmann opened the pull request:
~~Fully working / complete pre-templating version: #7481. This one here is an attempt to rebase post-templating datapath.~~ -> All OK here as well.

Description:

This work adds a NAT engine in BPF which is working together with
Cilium's BPF-based connection tracker. This gets enabled when the
following is set: `--install-iptables-rules=false --masquerade=true`
In this case Cilium won't rely on iptables to perform the masquerading
but instead uses its own, more efficient BPF-based implementation.

Right now at the beginning this option is only made available for
direct routing mode with ipvlan, but will be lifted in subsequent
work for veth-based datapath as well. In combination with ipvlan,
this allows to use the more efficient L3 mode instead of relying
on L3S! The faster L3 mode is otherwise not possible to use in
combination with masquerading since it bypasses Netfilter hooks.
For the BPF-based implementation in this work it is not a problem
since tc-based hooks are not bypassed.

The NAT core is mostly stand-alone, and in this work integrated
into the bpf_netdev program, but could in future also be integrated
into others. The NAT code handles connections from containers/pods
but also host local ones. In case of the latter, it will integrate
mappings into the same NAT table to check for potential collisions
and if the case, then it will map them to a different port just like
any other connections from container/pods. For connection originating
from local host, they need to be passed through the BPF-based
connection tracker as well in order to manage lifetime of NAT
mappings. For this purpose, it reuses the global connection tracker.

NAT mappings piggy back on Cilium's CT garbage collector in order
to remove related mappings once connection is terminated. Both
support for IPv4 as well as IPv6 has been implemented and tested
in this work. L3 protos that are supported are: TCP, UDP, ICMP
and ICMPv6. For both ICMP protocols, echo request/reply are supported
and in case of collisions their request/reply identifier translated.
Other ICMP types are not implemented yet, for that we'd need to
parse deeper into the packet in order to avoid leaking addressing
information. This can be done in future work. The NAT engine first
tries to retain ports/identifier from the packet in order avoid L4
packet rewrites, but in case of collisions with existing mappings,
it will select a different one. In case of L3 for local host
connections, it avoids rewrites altogether if the tuple can be
retained.

The NAT engine works as well with older kernels since it doesn't
rely on recently added helpers. Instead, it uses a subset of what
Cilium already uses today. There has been queue/stack map added to
the kernel some time ago (https://lwn.net/Articles/768994/), with
the motivation of implementing NAT through this, but we found that
using this map is rather limiting as a design with the help of that
map would only allow for 64k NAT mappings, but nothing beyond that.
For the queue map case, theory of operation would have been: fill
queue map with individual, unique port values as entries, and upon
port assignment, take a unused port out of the map to avoid
collisions, and once connection is terminated, put it back into
queue map. Aside from the limited number of mappings, this also
has the issue that we're wasting quite some amount of memory since
we need to prealloc the map with 64k entries in total 6 times (TCP,
UDP, ICMP for IPv4 and IPv6 each). We instead rely on randomized
port allocation and in case of collisions we retry up to 16 times
before giving up. The prandom based allocation is uniformly
distributed and aside from overcoming the memory waste, we also
allow for colliding port mappings as long as the tuple itself is
still unique thus avoiding a limit of 64k we would have had with
queue map otherwise.

The Cilium daemon has two new commands for managing the NAT table
similarly as with the connection tracking tables:

  `cilium bpf nat list`  - List all current NAT mappings
  `cilium bpf nat flush` - Remove all NAT mappings

Both are quite useful for introspection and/or debugging. Example
invocation with BPF based NAT engine:

Start daemon with BPF-based masquerading:
```
  # ./daemon/cilium-agent --kvstore consul \
                          --kvstore-opt consul.address=127.0.0.1:8500 \
                          --datapath-mode=ipvlan \
                          --ipvlan-master-device=bond0 \
                          --tunnel=disabled \
                          --install-iptables-rules=false \      <-- Switch to BPF
                          --masquerade=true \                   <-- Masquerading
                          --auto-direct-node-routes=true \
                          --enable-ipv4=true \
                          --enable-ipv6=true \
                          --disable-envoy-version-check=true
```
Check Cilium operates in ipvlan L3 mode:
```
  # ip -d a
  [...]
  277: cilium_host@bond0: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
      link/ether 0c:c4:7a:86:29:f8 brd ff:ff:ff:ff:ff:ff promiscuity 0
      ipvlan  mode l3 bridge numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
      inet 10.155.0.1/32 scope link cilium_host
         valid_lft forever preferred_lft forever
      inet6 2604:1380:2000:3d00::3/128 scope global
         valid_lft forever preferred_lft forever
      inet6 fe80::cc4:7a00:e986:29f8/64 scope link
         valid_lft forever preferred_lft forever
```
List current NAT mappings (ssh to node itself):
```
  # cilium bpf nat list
  TCP OUT 147.75.83.155:22 -> 188.63.217.235:51372 XLATE_SRC 147.75.83.155:22 Created=5sec HostLocal=1
  TCP IN 188.63.217.235:51372 -> 147.75.83.155:22 XLATE_DST 147.75.83.155:22 Created=5sec HostLocal=1
```
Ping from container to outside world with ipvlan L3:
```
  # docker exec -ti server ping 1.1.1.1
  PING 1.1.1.1 (1.1.1.1): 56 data bytes
  64 bytes from 1.1.1.1: seq=0 ttl=58 time=1.037 ms
  64 bytes from 1.1.1.1: seq=1 ttl=58 time=1.100 ms
  [...]
```
Again, list new NAT mappings:
```
  # cilium bpf nat list
  ICMP IN 1.1.1.1:0 -> 147.75.83.155:35328 XLATE_DST 10.155.53.137:35328 Created=4sec HostLocal=0
  TCP OUT 147.75.83.155:22 -> 188.63.217.235:51372 XLATE_SRC 147.75.83.155:22 Created=27sec HostLocal=1
  ICMP OUT 10.155.53.137:35328 -> 1.1.1.1:0 XLATE_SRC 147.75.83.155:35328 Created=4sec HostLocal=0
  TCP IN 223.111.139.244:33154 -> 147.75.83.155:22 XLATE_DST 147.75.83.155:22 Created=17sec HostLocal=1
  TCP OUT 147.75.83.155:22 -> 223.111.139.244:33154 XLATE_SRC 147.75.83.155:22 Created=17sec HostLocal=1
  TCP IN 188.63.217.235:51372 -> 147.75.83.155:22 XLATE_DST 147.75.83.155:22 Created=27sec HostLocal=1
```
Test manually flushing NAT mappings (test/demo-only, not needed otherwise):
```
  # cilium bpf nat flush
  Flushed 8 entries from /sys/fs/bpf/tc/globals/cilium_snat_v4_external
  Flushed 0 entries from /sys/fs/bpf/tc/globals/cilium_snat_v6_external
```
Ping from local host to outside world:
```
  # ping6 ipv6.google.com
  PING ipv6.google.com(ams16s29-in-x0e.1e100.net (2a00:1450:400e:804::200e)) 56 data bytes
  64 bytes from ams16s29-in-x0e.1e100.net (2a00:1450:400e:804::200e): icmp_seq=1 ttl=56 time=1.33 ms
  64 bytes from ams16s29-in-x0e.1e100.net (2a00:1450:400e:804::200e): icmp_seq=2 ttl=56 time=1.29 ms
  64 bytes from ams16s29-in-x0e.1e100.net (2a00:1450:400e:804::200e): icmp_seq=3 ttl=56 time=1.41 ms
  [...]
```
List mappings again:
```
  # cilium bpf nat list
  [...]
  ICMPv6 OUT [2604:1380:2000:3d00::3]:20631 -> [2a00:1450:400e:804::200e]:0 XLATE_SRC [2604:1380:2000:3d00::3]:20631 Created=5sec HostLocal=1
  ICMPv6 IN [2a00:1450:400e:804::200e]:0 -> [2604:1380:2000:3d00::3]:20631 XLATE_DST [2604:1380:2000:3d00::3]:20631 Created=5sec HostLocal=1
```
Flush of CT triggers eviction of NAT mappings as well. This, of course,
works in case of explicit flush (demoed here) as well as automatic
garbage collection:
```
  # cilium bpf ct flush global
  [...]
```
List mappings after flush (only ssh to node itself remaining):
```
  # cilium bpf nat list
  TCP OUT 147.75.83.155:22 -> 188.63.217.235:51372 XLATE_SRC 147.75.83.155:22 Created=5sec HostLocal=1
  TCP IN 188.63.217.235:51372 -> 147.75.83.155:22 XLATE_DST 147.75.83.155:22 Created=5sec HostLocal=1
```
Documentation under ipvlan has been updated, once we enable this
setting also for veth-based direct routing, then we'll add its own
getting started section as well.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7527)
<!-- Reviewable:end -->

tgraf commented on bpf/lib/nat.h:
```diff
@@ -0,0 +1,866 @@
+	}
+
+#pragma unroll
+	for (retries = 0; retries < 16; retries++) {
```
Define so we can change `16` via daemon flag?

borkmann replied:
Makes sense, I'll integrate this into daemon to emit a define, so we can tweak it from there when we see need.

tgraf commented on cilium/cmd/bpf_nat_flush.go:
```diff
@@ -0,0 +1,61 @@
+		}
+		if err != nil {
+			if err == os.ErrNotExist {
+				Fatalf("Unable to open %s: %s: please try using \"cilium bpf nat flush\"", path, err)
```
The suggestion in the error message seems weird

borkmann replied:
Agree, copy-paste issue from CT flush. Removed.

borkmann committed:
cilium: add cilium bpf nat cli commands

Add some CLI commands for inspecting and debugging BPF NAT.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ianvernon commented on pkg/tuple/ipv4.go:
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
```
Please add `pkg/tuple` to be owned by `cilium/bpf` in CODEOWNERS.

borkmann replied:
Ok, done.

borkmann committed:
maps, ctmap: refactor keys into tuple for later reuse outside of ct

They will later on be resued for the NAT map as well, just the
values will be different.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: initial bpf-based masquerading support

This work adds a NAT engine in BPF which is working together with
Cilium's BPF-based connection tracker. This gets enabled when the
following is set: --install-iptables-rules=false --masquerade=true
In this case Cilium won't rely on iptables to perform the masquerading
but instead uses its own, more efficient BPF-based implementation.

Right now at the beginning this option is only made available for
direct routing mode with ipvlan, but will be lifted in subsequent
work for veth-based datapath as well. In combination with ipvlan,
this allows to use the more efficient L3 mode instead of relying
on L3S! The faster L3 mode is otherwise not possible to use in
combination with masquerading since it bypasses Netfilter hooks.
For the BPF-based implementation in this work it is not a problem
since tc-based hooks are not bypassed.

The NAT core is mostly stand-alone, and in this work integrated
into the bpf_netdev program, but could in future also be integrated
into others. The NAT code handles connections from containers/pods
but also host local ones. In case of the latter, it will integrate
mappings into the same NAT table to check for potential collisions
and if the case, then it will map them to a different port just like
any other connections from container/pods. For connection originating
from local host, they need to be passed through the BPF-based
connection tracker as well in order to manage lifetime of NAT
mappings. For this purpose, it reuses the global connection tracker.

NAT mappings piggy back on Cilium's CT garbage collector in order
to remove related mappings once connection is terminated. Both
support for IPv4 as well as IPv6 has been implemented and tested
in this work. L3 protos that are supported are: TCP, UDP, ICMP
and ICMPv6. For both ICMP protocols, echo request/reply are supported
and in case of collisions their request/reply identifier translated.
Other ICMP types are not implemented yet, for that we'd need to
parse deeper into the packet in order to avoid leaking addressing
information. This can be done in future work. The NAT engine first
tries to retain ports/identifier from the packet in order avoid L4
packet rewrites, but in case of collisions with existing mappings,
it will select a different one. In case of L3 for local host
connections, it avoids rewrites altogether if the tuple can be
retained.

The NAT engine works as well with older kernels since it doesn't
rely on recently added helpers. Instead, it uses a subset of what
Cilium already uses today. There has been queue/stack map added to
the kernel some time ago (https://lwn.net/Articles/768994/), with
the motivation of implementing NAT through this, but we found that
using this map is rather limiting as a design with the help of that
map would only allow for 64k NAT mappings, but nothing beyond that.
For the queue map case, theory of operation would have been: fill
queue map with individual, unique port values as entries, and upon
port assignment, take a unused port out of the map to avoid
collisions, and once connection is terminated, put it back into
queue map. Aside from the limited number of mappings, this also
has the issue that we're wasting quite some amount of memory since
we need to prealloc the map with 64k entries in total 6 times (TCP,
UDP, ICMP for IPv4 and IPv6 each). We instead rely on randomized
port allocation and in case of collisions we retry up to 16 times
before giving up. The prandom based allocation is uniformly
distributed and aside from overcoming the memory waste, we also
allow for colliding port mappings as long as the tuple itself is
still unique thus avoiding a limit of 64k we would have had with
queue map otherwise.

The Cilium daemon has two new commands for managing the NAT table
similarly as with the connection tracking tables:

  cilium bpf nat list  - List all current NAT mappings
  cilium bpf nat flush - Remove all NAT mappings

Both are quite useful for introspection and/or debugging. Example
invocation with BPF based NAT engine:

Start daemon with BPF-based masquerading:

  # ./daemon/cilium-agent --kvstore consul \
                          --kvstore-opt consul.address=127.0.0.1:8500 \
                          --datapath-mode=ipvlan \
                          --ipvlan-master-device=bond0 \
                          --tunnel=disabled \
                          --install-iptables-rules=false \      <-- Switch to BPF
                          --masquerade=true \                   <-- Masquerading
                          --auto-direct-node-routes=true \
                          --enable-ipv4=true \
                          --enable-ipv6=true \
                          --disable-envoy-version-check=true

Check Cilium operates in ipvlan L3 mode:

  # ip -d a
  [...]
  277: cilium_host@bond0: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
      link/ether 0c:c4:7a:86:29:f8 brd ff:ff:ff:ff:ff:ff promiscuity 0
      ipvlan  mode l3 bridge numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
      inet 10.155.0.1/32 scope link cilium_host
         valid_lft forever preferred_lft forever
      inet6 2604:1380:2000:3d00::3/128 scope global
         valid_lft forever preferred_lft forever
      inet6 fe80::cc4:7a00:e986:29f8/64 scope link
         valid_lft forever preferred_lft forever

List current NAT mappings (ssh to node itself):

  # cilium bpf nat list
  TCP OUT 147.75.83.155:22 -> 188.63.217.235:51372 XLATE_SRC 147.75.83.155:22 Created=5sec HostLocal=1
  TCP IN 188.63.217.235:51372 -> 147.75.83.155:22 XLATE_DST 147.75.83.155:22 Created=5sec HostLocal=1

Ping from container to outside world with ipvlan L3:

  # docker exec -ti server ping 1.1.1.1
  PING 1.1.1.1 (1.1.1.1): 56 data bytes
  64 bytes from 1.1.1.1: seq=0 ttl=58 time=1.037 ms
  64 bytes from 1.1.1.1: seq=1 ttl=58 time=1.100 ms
  [...]

Again, list new NAT mappings:

  # cilium bpf nat list
  ICMP IN 1.1.1.1:0 -> 147.75.83.155:35328 XLATE_DST 10.155.53.137:35328 Created=4sec HostLocal=0
  TCP OUT 147.75.83.155:22 -> 188.63.217.235:51372 XLATE_SRC 147.75.83.155:22 Created=27sec HostLocal=1
  ICMP OUT 10.155.53.137:35328 -> 1.1.1.1:0 XLATE_SRC 147.75.83.155:35328 Created=4sec HostLocal=0
  TCP IN 223.111.139.244:33154 -> 147.75.83.155:22 XLATE_DST 147.75.83.155:22 Created=17sec HostLocal=1
  TCP OUT 147.75.83.155:22 -> 223.111.139.244:33154 XLATE_SRC 147.75.83.155:22 Created=17sec HostLocal=1
  TCP IN 188.63.217.235:51372 -> 147.75.83.155:22 XLATE_DST 147.75.83.155:22 Created=27sec HostLocal=1

Test manually flushing NAT mappings (test/demo-only, not needed otherwise):

  # cilium bpf nat flush
  Flushed 8 entries from /sys/fs/bpf/tc/globals/cilium_snat_v4_external
  Flushed 0 entries from /sys/fs/bpf/tc/globals/cilium_snat_v6_external

Ping from local host to outside world:

  # ping6 ipv6.google.com
  PING ipv6.google.com(ams16s29-in-x0e.1e100.net (2a00:1450:400e:804::200e)) 56 data bytes
  64 bytes from ams16s29-in-x0e.1e100.net (2a00:1450:400e:804::200e): icmp_seq=1 ttl=56 time=1.33 ms
  64 bytes from ams16s29-in-x0e.1e100.net (2a00:1450:400e:804::200e): icmp_seq=2 ttl=56 time=1.29 ms
  64 bytes from ams16s29-in-x0e.1e100.net (2a00:1450:400e:804::200e): icmp_seq=3 ttl=56 time=1.41 ms
  [...]

List mappings again:

  # cilium bpf nat list
  [...]
  ICMPv6 OUT [2604:1380:2000:3d00::3]:20631 -> [2a00:1450:400e:804::200e]:0 XLATE_SRC [2604:1380:2000:3d00::3]:20631 Created=5sec HostLocal=1
  ICMPv6 IN [2a00:1450:400e:804::200e]:0 -> [2604:1380:2000:3d00::3]:20631 XLATE_DST [2604:1380:2000:3d00::3]:20631 Created=5sec HostLocal=1

Flush of CT triggers eviction of NAT mappings as well. This, of course,
works in case of explicit flush (demoed here) as well as automatic
garbage collection:

  # cilium bpf ct flush global
  [...]

List mappings after flush (only ssh to node itself remaining):

  # cilium bpf nat list
  TCP OUT 147.75.83.155:22 -> 188.63.217.235:51372 XLATE_SRC 147.75.83.155:22 Created=5sec HostLocal=1
  TCP IN 188.63.217.235:51372 -> 147.75.83.155:22 XLATE_DST 147.75.83.155:22 Created=5sec HostLocal=1

Documentation under ipvlan has been updated, once we enable this
setting also for veth-based direct routing, then we'll add its own
getting started section as well.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
maps, natmap: add a new NAT map to access mappings from daemon

Add representation of NAT map in order to i) dump and flush the map
and ii) piggy back on garbage collection from ctmap.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: refactor ct maps into lib/{nat,conntrack_map}.h

We need to reuse the connection tracker from NAT as well, therefore
move the map definitions out of bpf_lxc. Also, take out the ct delete
helpers since we'll add NAT deletion logic in there as well.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: track icmp echo ids in conntrack entries

This is needed such that the conntrack garbage collection can later
also properly flush NAT mappings where we track and rewrite the echo
request/reply IDs instead of ports. Otherwise we would need separate
garbage collection runs for ICMP in NAT table only, which would be
ugly. There should be no breakage from CT side, since i) echo
req/replies are short lived, entries will be recreated and we also add
icmp related entries on top.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on bpf/lib/nat.h:
```diff
@@ -0,0 +1,872 @@
+	if (ret > 0)
+		return TC_ACT_OK;
+	if (ret < 0)
+		return DROP_NAT_NO_MAPPING;
```
This hides issues like connection tracker full, should this just return `ret`? (Same for IPv6)

borkmann replied:
Yes, makes sense.

joestringer commented on bpf/node_config.h:
```diff
@@ -82,3 +103,14 @@ DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x
 #define ENABLE_IPSEC
+
+#ifdef ENABLE_MASQUERADE
+#define CT_MAP_TCP6 test_cilium_ct_tcp6_65535
```
Remind me again how did you get around this overriding per-endpoint CT map selection (local/global)? Is it just because `ENABLE_MASQUERADE` protects this section and from endpoints BPF programs, `ENABLE_MASQUERADE` is never set?

borkmann replied:
`ENABLE_MASQUERADE` should never be set for endpoints. That said, it might have worked just like this because it was using global table in both cases (and in lxc case we included node_config.h). Will double check tomorrow.

joestringer commented on pkg/maps/nat/nat.go:
```diff
@@ -0,0 +1,264 @@
+			sizeKey,
+			sizeVal,
+			MaxEntries,
+			bpf.BPF_F_NO_PREALLOC, 0,
```
Usually the map preallocation is governed by the global preallocation daemon option, see things like `GetPreAllocateMapFlags()`. Is there a particular reason not to do the same here?

borkmann replied:
I could theoretically reuse them, my thinking was to not do it here but have the NAT table potentially large, but w/o having to carry this footprint all the time.

joestringer commented on bpf/lib/nat.h:
```diff
@@ -0,0 +1,872 @@
+	return 0;
+}
+
+static __always_inline int snat_v4_handle_mapping(struct __sk_buff *skb,
```
This function would really benefit from having a brief comment above describing the inputs and expected output possibilities.

In particular, the case where there's no corresponding NAT state and it's ingress where we just pass the packet (presumably so that the stack can just attempt to receive it directly) is a bit difficult to decipher.

borkmann replied:
Yeah, will do. (And yes, it's for receiving directly.)

joestringer reviewed (approved):
This is awesome.

Given the one giant commit, to review individual changes I ended up breaking it down into a bunch of separate commits, then I've placed some basic suggestions into commits adjacent to the sets of changes that I split out (TL;DR nothing major):

https://github.com/joestringer/cilium/commits/review/bpf-based-masquerade-rebased

This branch does not fold in any improvements regarding comments below.

Previously, in this area we were concerned about potential clashes between local applications reaching out to the same destination as traffic that is received to the node and masqueraded/forwarded to another node. If I follow correctly, this PR proposes to route all such local traffic into BPF so that we do NAT for that traffic using the same logic as well (using the Cilium internal IP), which means that all traffic leaving the node to peer nodes that Cilium manages is NAT'd using the same implementation, and furthermore replies are received back into the BPF logic first, then un-NAT'd and routed back to their original sources. Hence we no longer have the PAT collision issue. Is that right? 

I added a few notes below. Nothing really stands out as likely to cause problems, but it's also pretty easy to miss things when the code just looks like it's doing all the right things :-). It'd be good to file an issue or somehow track adding end-to-end tests of this functionality.
